Handle non-integer error codes as exit code 1 (#2115)

nkabardin opened the pull request:
Fixes #2115 

When there is an error in require() in configuration file, cli gets an error code which is not actually integer, so passing it to process.exit is not a good idea.

Now value is getting checked if it's a number or not, and if it's not a number, it's replaced with plain `1`.

System test added that runs cli in ensures that exit code != 0

nkabardin commented:
@davidtheclark thanks!

something is frozen on ci though. Can't understand what's happening.

nkabardin commented:
Checked log on Travis, and there might be issues with running this system test on jenkins — any ideas what's happening there?

ntwb commented:
@nkabardin I've restarted the CI tests so lets see what happens...

ntwb commented:
Coveralls is failing in both `5569.1` and `5569.2` jobs even though `#5569.1` [is reporting](https://travis-ci.org/stylelint/stylelint/jobs/179432287) as `passed` and `#5569.2` is failing with a Jest test timing out.

nkabardin commented:
So I guess it's something related to `child_process.spawn`, are there any CI-related limitations for that?

nkabardin commented:
Playing around with error logging in system test, let's see what CI will tell us...

nkabardin commented:
did several attempts trying to log out the problems with running stuff, added three `console.log` to test, and didn't catch the problem, nothing logged.

Any ideas?

nkabardin commented:
Rebased branch over current master (so, support babel drop and no need to use babel-node to run cli) and also  and removed all debug stuff and it's currently ready and well tested, and surprisingly issues on CI are also solved, so we got green build there, so yay!

ntwb commented:
LGTM, though let's get another review from @davidtheclark 👍

ntwb commented:
Coverage tests were failing if less than 95%, I've dropped that value to 90%

So ignore the Coveralls failure here in this PR (I couldn't find a way to restart the job)

davidtheclark commented on system-tests/004/004.test.js:
```diff
@@ -0,0 +1,24 @@
+  const localPath = path.resolve(__dirname)
+  const cliPath = path.join(localPath, "../../lib/cli.js")
+
+  const cliProcess = spawn("node", [ cliPath, `${localPath}/*.css` ], { cwd: localPath })
```
Could you add an `'error'` listener to this process?

nkabardin replied:
done

evilebottnawi commented on system-tests/004/004.test.js:
```diff
@@ -0,0 +1,28 @@
+  const cliProcess = spawn("node", [ cliPath, `${localPath}/*.css` ], { cwd: localPath })
+
+  cliProcess.on("error", function (error) {
+    console.log("error running cli:", error) // eslint-disable-line no-console
```
I think we should failed test if have error in `error` event

nkabardin replied:
it fails anyway if there is an error, it just added logging for that.
Couldn't reproduce a case when it didn't fail on error

evilebottnawi replied:
@nkabardin we can use `mock` for reproduce error, we got non zero code (thoretically :smile: ) and test failed, lgtm

nkabardin replied:
I didn't get it, current code looks good to you, or it will look good to you when some mocking will be added? :)

evilebottnawi replied:
@nkabardin no, all ok

davidtheclark commented:
Thanks again @nkabardin!

davidtheclark commented:
Changelog entry

> -   Fixed: bug where some Node errors in special cases did not cause the CLI to exit with a non-zero code ([#2140](https://github.com/stylelint/stylelint/pull/2140))

nkabardin commented:
@davidtheclark glad I was able to help the project! thanks for `stylelint` ❤️
